tracer/src/Datadog.Trace/ClrProfiler/ServerlessInstrumentation/LambdaHandler.cs

```diff
@@ -29,36 +29,90 @@ internal LambdaHandler(string? handlerName)
+        var handlerMethod = handlerType.GetMethod(methodName);
         if (handlerMethod is null)
         {
-            throw new Exception($"Could not find handler method for {handlerName}");
+            ThrowHelper.ThrowArgumentException($"Could not find handler method {handlerName} in type {handlerType}");
+        }
+
+        // Generic methods aren't supported in AWS Lambda
+        // https://docs.aws.amazon.com/lambda/latest/dg/csharp-handler.html#csharp-handler-restrictions
+        if (handlerMethod.IsGenericMethod || handlerMethod.IsGenericMethodDefinition)
+        {
+            ThrowHelper.ThrowArgumentException($"Unable to instrument generic method {handlerMethod.Name} declared on {handlerMethod.DeclaringType}");
         }
 
-        if (handlerMethod.IsGenericMethod || handlerMethod.DeclaringType?.IsGenericType == true)
+        var declaringType = handlerMethod.DeclaringType;
+
+        // We should never be invoking generic _definitions_ (i.e. open generics)
+        // because the handlerType should always be a specialised type
+        if (declaringType?.IsGenericTypeDefinition == true)
         {
-            throw new Exception($"Unable to instrument generic handler method {handlerMethod.Name} declared on {handlerMethod.DeclaringType}");
+            ThrowHelper.ThrowArgumentException($"Unable to instrument generic method {handlerMethod.Name} declared on generic type definition {declaringType}");
         }
 
-        // The method body may be in a different type, e.g. a base type
-        // In that case we need the FullType to point to the base type
-        FullType = handlerMethod.DeclaringType?.FullName ?? handlerTokens[1];
-        // If the handlerType == the declaring type, skip calling Assembly.GetName and use handler token directly
-        Assembly = handlerType == handlerMethod.DeclaringType
-                       ? handlerTokens[0]
-                       : handlerMethod.DeclaringType?.Assembly.GetName().Name ?? handlerTokens[0];
+        if (declaringType is not null
+         && declaringType != handlerType
+         && declaringType.IsGenericType)
+        {
+            // Handler is declared on a base type, which is generic
+            // We need to instrument the method in the generic type, not the specialized version
+            var genericTypeDefinition = declaringType.GetGenericTypeDefinition();
+
+            // The handlerMethod _currently_ points to the specialised version of the method
+            // But we need to get the non-specialised version
+            // I can't find a way to get this directly from genericDefinition or handlerMethod,
+            // so we search for it directly instead
+            var genericMethod = genericTypeDefinition.GetMethod(handlerMethod.Name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly | BindingFlags.Instance | BindingFlags.Static);
+            if (genericMethod is null || !genericMethod.ContainsGenericParameters)
```

Comment:
Is there a scenario where we need to instrument a non-generic method inside of a generic type? Like if the method had took no arguments and returned void? Or if it only had built-in types like `string` in the arguments or return types?

Reply:
We definitely do want to support that, and the current code already does! Interestingly, this interacts somewhat with the previous example I think. 

For example if we have a type like this:

```csharp
public class DerivedImplementation : GenericBase<CustomInput, string>
{
}

public abstract class GenericBase<TRequest, TResponse>
{
    public Task NonGenericFunction(string arg1, ILambdaContext arg2) => Task.CompletedTask;
}
```

then `genericMethod.ContainsGenericParameters` is _still_ `true`, even though it's neither a generic method, nor contains any parameters that are generic. Specifically:

```csharp
ContainsGenericParameters = true;
GenericParameterCount = 0;
IsGenericMethod = false;
GetGenericArguments() => Array.Empty<Type>();
```

So `ContainsGenericParameters` relates to the fact it's a method _in_ a generic type. 

I added an extra unit test case to confirm it works as expected.

Confusing stuff 😅

Reply:
Thank you for the explanation! Sounds good to me, and thank you for the additional unit test case